### DISCUSSION TITLE
How to run laernlib

### DISCUSSION THREAD

**Discussion by 5angjun** [LearnLib/learnlib#131] [Q&A]
Hello, I'm sangjun who is very interested in your project.

I have at a setup stage for using learnlib.

When i execute this commmand, i have a problem as you can see below.
`mvn clean install`
![image](https://github.com/user-attachments/assets/f5ed3dea-fcaa-4759-92e2-bb6ad66308fb)

Is there any specific instruction for building and testing learnlib's learning algorithm?

For more detail, this is my environment and instruction
```
WSL ubuntu 18.0.4LTS
  286  git clone https://github.com/LearnLib/learnlib.git
  287  cd learnlib/
  288  ls
  289  git checkout release
  290  mvn clean install
```

**Comment by mtf90** [comment]
Dear @5angjun, can you maybe post the actual test failure message before the reactor summary? Just from knowing that

> There are test failures.

I cannot really see what is going wrong.


At the same time, if the build only fails at the learnlib-examples (the last module), this means that all of the other modules built without any issues, so you should be able to use them anyway.

However, I would suggest to not `mvn install` the versions from the release branch because this will override the public releases in your local Maven repository. Especially if you start to modify your checkout, this could result in very hard-to-debug issues. Either use a special version (such as the SNAPSHOT versions on the develop branch) or stick to the pre-built releases in the Maven central repository (you can include the LearnLib binaries in any Maven-aware build tool without having to explicitly build them yourself).

**Comment by 5angjun** [comment]
I work with released version of learnlib.(https://github.com/LearnLib/learnlib/releases)
But there is other question.

1. How can i apply it to toy SUT?
-> is there any example for real binary?
-> All of the example is written with newDFA(), but i want to write something real binary and write mapper for binary.
-> Written with newDFA() model is unfeasible with my real input. And i cannot execute that model in my hand. So it not that much understandable.
-> So there is any example with mapper and real toy binaries?
![스크린샷, 2024-09-02 00-01-44](https://github.com/user-attachments/assets/14180b81-a6f5-464d-8651-5b5e599fa786)

Thanks!

**Reply by mtf90** [reply]
If you only work with stable releases of LearnLib, you would typically setup your own Maven or Gradle project (i.e., with custom groupId, etc.) with LearnLib as a dependency and not import it in IntelliJ directly because the class loading may change depending on whether you run your code from the IDE or via Maven. If you know what you are doing, everything is fine. It is just meant as an advice to prevent potential head-scratchers in the future.


As for integrating external systems, your main task is to implement a `MembershipOracle<I, Boolean>` (or the simplified `SingleQueryOracle` if you don't care about parallelism) and pass this implementation to both the learning algorithm and the equivalence oracle. For example, [`Example2`](https://github.com/LearnLib/learnlib/blob/develop/examples/src/main/java/de/learnlib/example/Example2.java) shows how to query a Java class with the use of reflection. Although, note that here, a `SUL`-based interface is used because the example learns a Mealy machine. A `DFA` setup is typically simpler, because you only need to answer whether a query is accepted or not. How you implement the oracle highly depends on your toy example (e.g., sending requests over the network, invoking a pre-compiled binary, etc.).

**Comment by 5angjun** [comment]
Okay thanks for your helping.

I think i need to inspect and research more example code again.
After that, i will re-open the issue if i need more help or information.

Thanks you Markus.
